### PR TITLE
fix(vim.system): resolve executable paths on windows

### LINKED ISSUE #31107: vim.system does not respect executable file extensions for MS Windows
### Problem

For MS Windows there are multiple file extensions for executable files, like

- exe
- bat
- cmd
- com

These executable will be automatically found by the Windows system itself, like:

```sh
$ where winrm

C:\Windows\System32\winrm.cmd
C:\Windows\System32\winrm.vbs
```

the first one found in `%PATH%` or local directory will be run.

Normal "vim" implementation also respect these correctly (also when called from lua):

```vim
: lua print(vim.fn.executable("winrm"))

1
```

or

```vim
:lua print(vim.fn.system({"winrm"}))

Befehlszeilenprogramm zur Windows-Remoteverwaltung

Die Windows-Remoteverwaltung (WinRM) ist die Microsoft-Implementierung 
des WS-Verwaltungsprotokolls, das eine sichere Kommunikation 
...
```

BUT using `vim.system` there will be a self implemented check for the executable file which does not correctly test for all valid extensions (at least not for `.cmd`) :

```vim
:lua print(vim.inspect(vim.system({"winrm"})))

E5108: Error executing lua ...gram Files/Neovim/share/nvim/runtime/lua/vim/_system.lua:244: ENOENT: no such file or directory
stack traceback:
	[C]: in function 'error'
	...gram Files/Neovim/share/nvim/runtime/lua/vim/_system.lua:244: in function 'spawn'
	...gram Files/Neovim/share/nvim/runtime/lua/vim/_system.lua:335: in function 'system'
	[string ":lua"]:1: in main chunk
```

while giving the full executable with extensions runs fine:

```vim
:lua print(vim.inspect(vim.system({"winrm.cmd"})))

{
  _state = {
    cmd = <1>{ "winrm.cmd" },
    done = false,
    handle = <userdata 1>,
    pid = 47000,
    stderr = <userdata 2>,
    stderr_data = {},
    stdout = <userdata 3>,
    stdout_data = {}
  },
  cmd = <table 1>,
  pid = 47000,
  <metatable> = {
    __index = {
      _timeout = <function 1>,
      is_closing = <function 2>,
      kill = <function 3>,
      wait = <function 4>,
      write = <function 5>
    }
  }
}
```

Also the vim function `vim.fn.exepath` behaves normal and gives you the correct path:

```vim
:lua print(vim.fn.exepath("winrm"))

C:\Windows\system32\winrm.CMD
```

### Steps to reproduce

see above

### Expected behavior

All default executable extensions must be respected.

### Nvim version (nvim -v)

v0.11.0-dev-226+g721551210

### Vim (not Nvim) behaves the same?

n/a

### Operating system/version

MS Windows 11

### Terminal name/version

wezterm 20241007-103714-ed430415

### $TERM environment variable

xterm-256color

### Installation

repo

### PR DESCRIPTION
**Pull request by lewis6991**
Fixes #31107

### CODE REVIEW

**Comment by lewis6991** [comment]
Not sure if I can write a test for this. @Grueslayer do you have any ideas. Here's an example of a current test:

```lua
      it('handle input', function()
        eq('hellocat', system({ 'cat' }, { stdin = 'hellocat', text = true }).stdout)
      end)
```

**Comment by gorillamoe** [comment]
We (@Grueslayer and me) had this issue with nodejs and `npm` not being found on a windows systems. Idk if you have windows specific tests.

For ref: https://github.com/mistweaverco/kulala.nvim/issues/304

So our problematic code on windows was something like `vim.system({"npm", "-v"})`.

Idk about other programs that might have problems with that call, but npm sure does.

**Comment by lewis6991** [comment]
> Feels hacky, but you could put a awesomeapp.bat in C:\windows\system32\ and a awesomeapp in /usr/bin and just echo 1 and test that both these calls succeed. 🙈

Good idea this should work well since we can just create a temp dir and add that to PATH.

**Comment by Grueslayer** [comment]
@lewis6991 As in my description, winrm.cmd should be available in default windows installation:

**Comment by lewis6991** [comment]
Preferably the test should spawn something that is benign.

File: runtime/lua/vim/_system.lua [hunk 1]
```diff
@@ -230,6 +230,8 @@ local function default_handler(stream, text, bucket)
   end
 end
 
+local is_win = vim.fn.has('win32') == 1
```
Comment (gorillamoe) [inline runtime/lua/vim/_system.lua:233]: Is there also win64?

Comment (lewis6991) [reply runtime/lua/vim/_system.lua:233]: No. I think this just means all windows.

Comment (Grueslayer) [reply runtime/lua/vim/_system.lua:233]: There is a win64 but win32 includes 32 and 64 bit, so everything is fine

**Comment by mfussenegger** [comment]
Might be worth setting hide too if .cmd is used simplicity. See https://github.com/mfussenegger/nvim-lint/pull/649

**Comment by lewis6991** [comment]
You mean special case `cmd.exe`?

**Comment by lewis6991** [comment]
I'm having problems with the test:

```lua
    it('can resolve windows command extentions.', function()
      t.write_file('test.bat', '')
      system_sync({ 'chmod', '+x', 'test.bat' })
      system_sync({ './test' })
    end)
```
is erroring with:

```

I don't know, use or have access to a windows machine so I'm a bit stuck here.

@dundargoc HALP!
test\functional\testnvim.lua:950: Error executing lua: D:\a\neovim\neovim\runtime/lua/vim/_system.lua:253: EFTYPE: inappropriate file type or format
stack traceback:
	[C]: in function 'error'
	D:\a\neovim\neovim\runtime/lua/vim/_system.lua:253: in function 'spawn'
	D:\a\neovim\neovim\runtime/lua/vim/_system.lua:342: in function 'system'
	test/functional\lua\system_spec.lua:10: in function 'f'
	test\functional\testnvim.lua:850: in function 'handler'
	[string "<nvim>"]:1: in main chunk

stack traceback:
	test\functional\testnvim.lua:950: in function 'system_sync'
	test/functional\lua\system_spec.lua:120: in function <test/functional\lua\system_spec.lua:117>
```

**Comment by dundargoc** [comment]
Home in 2 hours, I'll check then if it's nit solved until then

**Comment by Grueslayer** [comment]
@lewis6991 You need at least have a content in your BAT file which could be executed, like

```
@echo off
echo Hello World
```
with an empty file you'll get an error.

I don't know your test framework but maybe you need to use Windows backslashes for path separation.

Let's assume you created a file `test.bat` with the above content in the current directory (which is e.g. `c:\temp\`) and use the following vim commands:

```vim
:cd c:\Temp\
:lua print(vim.inspect(vim.system({ vim.fn.exepath(".\\test") })))
```

everything runs fine (as your fix does the same, it should run without the extra call in the arguments)

**Pull request merged**